### PR TITLE
fix: add validation for programStage.program

### PR DESCRIPTION
**Pull request by vietnguyen**
https://jira.dhis2.org/browse/DHIS2-12129

## Issue
- Currently user can create a ProgramStage which does not have reference to any Program using `api/programStages` or `api/metadata`

## Fix
- This PR will add validation for programStage.program association for importService.
- The validation will throw error `ErrorCode.E4053` if  programStage.program is null AND programStage is not referenced from any programs in same payload.
- The fix will affect `/api/programStages` and `/api/metadata` endpoints.
- After the fix
  - `api/programStages` will always require programStage to have a reference to program.
  - `api/metadata` will require programStage to have a reference to program OR it should be referenced from a program in same payload.
- I also modified some e2e and unit tests because those tests violate the new validation.

## Test
- Can manual test using simple payload like below, remove the program references to check if the validation work.

POST to `api/metadata`
```
{
  "programs":
  [
    {
      "id": "lw62xvyB26a",
      "name": "test",
      "shortName": "test",
      "programType": "WITH_REGISTRATION",
      "programStages":
      [
        {
          "id": "B8wWHAh1jgH"
        }
      ]
    }
  ],
  "programStages":
  [
    {
      "id": "B8wWHAh1jgH",
      "name": "test"
    }
  ]
}
```

POST to `api/programStages`
```
{
  "name": "te2st123",
  "program":
  {
    "id": "r641fNlNVD1"
  }
}
```

### CODE REVIEW

File: dhis-2/dhis-e2e-test/src/main/java/org/hisp/dhis/actions/metadata/ProgramActions.java [hunk 1]
```diff
@@ -131,22 +131,31 @@ public ApiResponse addProgramStage( String programId, String programStageId )
...
-            .post( new JsonObjectBuilder().addProperty( "name", name ).addProperty( "publicAccess", "rwrw----" ).build() );
+            .post( new JsonObjectBuilder().addProperty( "name", programStageName ).
+                addObject( "program" , new JsonObjectBuilder().addProperty( "id", programId ))
+                .addProperty( "publicAccess", "rwrw----" ).build() );
         response.validate().statusCode( Matchers.is( Matchers.oneOf( 201, 200 ) ) );
 
         return response.extractUid();
     }
 
+//    public String createProgramStage( String name, String programId )
```
Comment (larshelge) [inline dhis-2/dhis-e2e-test/src/main/java/org/hisp/dhis/actions/metadata/ProgramActions.java:148]: Remove commented code.

**Comment by larshelge** [comment]
Tests are failing @vietnguyen .

**Comment by vietnguyen** [comment]
> Tests are failing @vietnguyen .

Fixed now

**Review by jbee — APPROVED** [review]
Looks good.

File: dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/metadata/metadata_import/MetadataImportBasedOnSchemasTest.java [hunk 2]
```diff
@@ -104,7 +104,8 @@ public void postBasedOnSchema( String endpoint, String schema )
             "messageConversations",
             "users",
             "organisationUnitLevels",
-            "programRuleActions" ); // blacklisted because contains
+            "programRuleActions",
+            "programStages" ); // blacklisted because contains
```
Comment (vilkg) [inline dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/metadata/metadata_import/MetadataImportBasedOnSchemasTest.java:108]: could we mark program as required in schemas? Then we should need to blacklist this

Comment (vilkg) [reply dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/metadata/metadata_import/MetadataImportBasedOnSchemasTest.java:108]: ah, I see! LGTM then 👍

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ErrorCode.java [hunk 3]
```diff
@@ -200,6 +200,7 @@
     /* ProgramRuleVariable validation */
     E4051( "A program rule variable with name `{0}` and program uid `{1}` already exists" ),
     E4052( "For program rule variable with name `{0}` following keywords are forbidden : and , or , not" ),
+    E4053( "ProgramStage `{0}` does not have reference to any Program." ),
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ErrorCode.java:203]: We could rephrase a bit to:

    Program stage `{0}` does not reference a program

**Review by mortenoh — COMMENTED** [review]
There a lot of `Code Scanning` issues.. did you have a look at them?

**Comment by vietnguyen** [comment]
> There a lot of `Code Scanning` issues.. did you have a look at them?

Those are not related to my changes. 
All issues gone after merged with master branch.

**Review by larshelge — APPROVED** [review]

**Pull request merged**